contracts/TerminalV1Rescue.sol

```diff
@@ -0,0 +1,204 @@
+        _fundingCycles != IFundingCycles(address(0)) &&
+        _ticketBooth != ITicketBooth(address(0)) &&
+        _terminalDirectory != ITerminalDirectory(address(0)) &&
+        _owner != address(0),
+      'zero address'
+    );
+    projects = _projects;
+    fundingCycles = _fundingCycles;
+    ticketBooth = _ticketBooth;
+    terminalDirectory = _terminalDirectory;
+
+    transferOwnership(_owner);
+  }
+
+  /** 
+    @notice 
+    Allows a project to print tickets for a specified beneficiary.
+
+    @dev
+    Only a project's owner or a designated operator can print tickets.
+
+    @param _projectId The ID of the project to print tickets for.
+    @param _amount The amount of tickets to print.
+    @param _beneficiary The address to send the printed tickets to.
+    @param _memo A memo to leave with the printing.
+    @param _preferUnstakedTickets If there is a preference to unstake the printed tickets.
+  */
+  function printTickets(
+    uint256 _projectId,
+    uint256 _amount,
+    address _beneficiary,
+    string memory _memo,
+    bool _preferUnstakedTickets
+  )
+    external
+    override
+    requirePermission(projects.ownerOf(_projectId), _projectId, Operations.PrintTickets)
+  {
+    // Can't send to the zero address.
+    require(_beneficiary != address(0), 'zero address');
+
+    // Print the project's tickets for the beneficiary.
+    ticketBooth.print(_beneficiary, _projectId, _amount, _preferUnstakedTickets);
+
+    emit PrintTickets(_projectId, _beneficiary, _amount, _memo, msg.sender);
+  }
+
+  /** 
+    @notice
+    A function that burns the supply of the dead address for a project. 
+
+    @dev
+    Callable by anyone.
+
+    @param _projectId The ID of the project whose ETH is being resued.
+    @param _beneficiary The recipient of the resucued funds.
+    @param _amount The amount to rescue as a fixed point number.
+  */
+  function rescue(
+    uint256 _projectId,
+    address payable _beneficiary,
+    uint256 _amount
+  ) external override {
+    // Must be project owner.
+    require(msg.sender == projects.ownerOf(_projectId), 'not owner');
+
+    // Project must be allow listed.
+    require(rescueAllowed[_projectId], 'not allowed');
+
+    // Amount must have been specified.
+    require(_amount > 0, 'must be rescuing something');
+
+    // Get a referrence to the original balance.
+    uint256 _originalBalance = balanceOf[_projectId];
+
+    // Balance must exist.
+    require(_originalBalance >= _amount, 'not enough to rescue');
+
+    // Get a reference to the current funding cycle for the project.
+    FundingCycle memory _fundingCycle = fundingCycles.get(fundingCycles.latestIdOf(_projectId));
+
+    // Only one-time funding cycles are rescuable.
+    require(_fundingCycle.discountRate == 201, 'cant rescue');
+
+    // Send funds to beneficiary.
+    if (_originalBalance > 0) Address.sendValue(_beneficiary, _amount);
+
+    balanceOf[_projectId] = balanceOf[_projectId] - _amount;
```

Comment:
@mejango isn't there a reentry vulnerability here (a bad projectOwner draining the terminal)? Would swap ln144 and 146

Reply:
great great great eye